src/active_query.rs

```diff
@@ -48,7 +48,7 @@ pub(crate) struct ActiveQuery {
 
     /// Map from tracked struct keys (which include the hash + disambiguator) to their
     /// final id.
-    pub(crate) tracked_struct_ids: FxHashMap<KeyStruct, Id>,
+    pub(crate) tracked_struct_ids: FxHashMap<KeyStruct, DatabaseKeyIndex>,
```

Comment:
I'm not 100% sure if storing the `DatabaseKeyIndex` is necessary, but we now need a way to remove elements, and the `Id` itself isn't unique enough. Two tracked struct from different ingredients can both have the id 0. The ambiguity hasn't been a problem up to now because we only did lookups by key.

Reply:
Yeah, it seems necessary if we take this approach of filtering out afterwards. The alternative would be to have two hashmaps, one with the `tracked_struct_ids` from the previous revision, and one that you are building up with the new values. You would first check the old map and, if something is found, move it to the new map. Then you don't have to remove anything after the fact.

I thought about suggesting this but I think this retain-based design is probably mildly more efficient.

Reply:
> I thought about suggesting this but I think this retain-based design is probably mildly more efficient.

Yeah, I think so too. At least for functions that create many tracked structs. It also means that there's one extra allocation for every tracked function.